File: src/go/day2/puzzle1.go
package main

import "fmt"

func puzzle1() {
	fmt.Println(countSafe(Input, isSafe1))
}

func isSafe1(nums []int) bool {
	if len(nums) < 2 {
		return true
	}
	if nums[0] < nums[len(nums)-1] {
		return _isSafe1(nums, increasing)
	}
	return _isSafe1(nums, decreasing)
}

func _isSafe1(nums []int, direction int) bool {
	for i := 1; i < len(nums); i++ {
		diff := direction * (nums[i] - nums[i-1])
		if 1 <= diff && diff <= 3 {
			continue
		}
		return false
	}
	return true
}

File: src/python/day2/puzzle1.py
def safe1(arr):
    if len(arr) < 2:
        return True
    return _safe1(arr, 1) or _safe1(arr, -1)

def _safe1(arr, direction):
    for i in range(len(arr)-1):
        diff = direction*(arr[i+1]-arr[i])
        if diff < 1 or 3 < diff:
            return False
    return True

if __name__ == "__main__":
    print(safe1([1, 2, 3]))
    print(safe1([1, 2, 5]))
    print(safe1([1, 2, 6]))
    print(safe1([1, 2, 1]))
    print(safe1([3, 2, 1]))
    print(safe1([1, 2, 4, 3]))
    print(safe1([1, 2, 4, 10]))